### ISSUE TITLE
Rendering increasing numbers of points becomes quite slow

### ISSUE THREAD

**Issue by Sushisource** [ggez/ggez#127]
In my looking through the code I think this isn't so much specific to points, but rather just having lots and lots of polys being drawn (since the points are converted into rects, which are converted into polys).

In my case I'm rendering a cellular automata simulation, so there are many thousands of points being drawn. The slowdown starts to get pretty bad around the 6k mark, on my machine at least.

At least in the points case, I think this could be optimized by using GL_POINTS primitives, or whatever the backend-agnostic gfx-rs equivalent is, if such a thing exists.

Hopefully this is helpful info. I'm happy to contribute PRs, albeit probably pretty slowly.

**Comment by icefoxen** [comment]
I suspect the problem is less a matter of not using GL_POINTS (since iirc it is not well supported these days anyway) and more that each draw sends a separate openly draw call.

I believe the Right Answer is sprite batching, which lets you collect a large buffer of mostly-similar drawing instructions and then send them to the GPU all at once. There's a little precursor code to that but it's not connected to anything yet.

For cellular automata you might be well served just by creating a buffer with all your pixels in it, turning it into an Image with `Image::from_rgba8`, and displaying that. Example here: https://gist.github.com/icefoxen/ffde9dc776848c80e25656038e9248f9

**Comment by icefoxen** [comment]
However if you link some example code that displays the problem I can profile it and see what's actually going on.

**Comment by Sushisource** [comment]
Yeah, I'm by no means an OpenGL expert (or even amateur). Your solutions sound like they make perfect sense. I'll try out the Image based approach next.

Here's the game (well, very very start of something that maybe will be a game eventually) I'm doing the CA thing for. Forgive the mess 🤕. I didn't make a pared-down example but the amount of code here is pretty small anyway:
https://github.com/Sushisource/Rustlike

I just committed converting the code from glium to ggez. The slowdown while rendering many points should be pretty evident.

**Comment by Manghi** [comment]
I saw something very similar when trying to draw 0x8000 rectangles individually (also cellular automata-based 😋 ) under `ggez 3.0`, which was previously not seen under `ggez 2.x`. With some crude benchmarking, I narrowed it down `gfx::draw` calls. This might be a good place to look, however I did see on the `gfx` wiki page that they are not entirely focused on performance quite yet as they are trying to get a stable platform going. I agree that individually drawing that many rectangles is on the wrong side of the performance spectrum and that something like @icefoxen's suggestion would work (however I also am not an expert or even amateur and clearly this did not even come to mind).
